docs/middleware.md

```diff
@@ -0,0 +1,102 @@
+          "": "query"
+      }
+
+ The transformed request data is validated against the schema. Any validation errors are then transformed using
+ `options.rules.validationErrorsToResponse` before they are sent to the user.  The default format looks roughly like:
+
+     {
+       ok: false,
+       message: "The JSON you have provided is not valid.",
+       errors: {
+         field1: ["This field is required."]
+       }
+     }
+
+ The output of this middleware is itself expected to be valid according to a JSON schema, and is delivered using a
+ [`schemaHandler`](./handler.md).
+
+# Component Options
+
+The following component configuration options are supported:
+
+| Option              | Type     | Description |
+| ------------------- | -------- | ----------- |
+| `handlerGrades` | `Array` | An array of grade names that will be used in constructing our request handler. |
+| `method` | `String` | The method(s) the inner router will respond to.  These should be lowercase strings corresponding to the methods exposed by Express routers.  The default is to use the `POST` method, there are convenience grades for each method. |
+| `responseSchemaKey` | `String` | The schema key that [our handler](./handler.md) will use in constructing response headers. |
+| `responseSchemaUrl` | `String` |  A base URL where `responseSchemaKey` can be found. |
+| `rules.requestContentToValidate` | `Object` | The [rules to use in transforming](http://docs.fluidproject.org/infusion/development/ModelTransformationAPI.html#fluid-model-transformwithrules-source-rules-options-)
+the incoming data before validation (see above). |
+| `rules.validationErrorsToResponse` | `Object` | The [rules to use in transforming](http://docs.fluidproject.org/infusion/development/ModelTransformationAPI.html#fluid-model-transformwithrules-source-rules-options-)
+validation errors before they are sent to the user (see above). |
+| `schemaKey`         | `String` |  The key (also the filename) of the schema to be used for validation. |
+| `schemaPath`        | `String` | The path to a directory that contains a file matching `options.schemaKey`.  On the client-side, this is expected to be a full or relative URI.  On the server-side, this is expected to be a package-relative path such as `%gpii-handlebars/tests/schemas`. |
+
+# Invokers
+
+## `{middleware}.middleware(req, res, next)`
+
+* `req`: The [request object](http://expressjs.com/en/api.html#req) provided by Express, which wraps node's [`http.incomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage).
+* `res`: The [response object](http://expressjs.com/en/api.html#res) provided by Express, which wraps node's [`http.ServerResponse`](https://nodejs.org/api/http.html#http_class_http_serverresponse).
+* `next`: The next Express middleware or router function in the chain.
+* Returns: Nothing.
+
+This invoker fulfills the standard contract for a `gpii.express.middleware` component.  It examines the `req` content
```

Comment:
In my reading, that method signature is for middleware that _responds_ to errors, and not for middleware that _generates_ errors:

http://expressjs.com/en/guide/using-middleware.html#middleware.error-handling

Our middleware interrupts the conversation by neglecting to call the `next` function it is passed, as outlined (in reverse) in [the documentation](http://expressjs.com/en/guide/writing-middleware.html):

> If the current middleware function does not end the request-response cycle, it must call next() to pass control to the next middleware function. Otherwise, the request will be left hanging.

Having our middleware call `next(error)` is also an option here.  It would however mean that we'd have to wire in our own error-handling middleware and ensure that it is always "downstream" of this middleware.  Otherwise we would not have the same level of control over the response sent to the user.

I would suggest keeping the current implementation the way it is unless we have another use case for writing our own error-handling middleware or some more compelling reason why simply choosing to end the conversation and respond directly is not acceptable.